test/lit/passes/heap-store-optimization.wast

```diff
@@ -876,5 +884,751 @@
       (local.get $ref)
     )
   )
-)
 
+  ;; CHECK:      (func $control-flow-in-set-value-safe (type $4) (result i32)
+  ;; CHECK-NEXT:  (local $ref (ref null $struct))
+  ;; CHECK-NEXT:  (block
+  ;; CHECK-NEXT:   (local.set $ref
+  ;; CHECK-NEXT:    (struct.new $struct
+  ;; CHECK-NEXT:     (if (result i32)
+  ;; CHECK-NEXT:      (i32.const 1)
+  ;; CHECK-NEXT:      (then
+  ;; CHECK-NEXT:       (i32.const 1337)
+  ;; CHECK-NEXT:      )
+  ;; CHECK-NEXT:      (else
+  ;; CHECK-NEXT:       (i32.const 42)
+  ;; CHECK-NEXT:      )
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (struct.get $struct 0
+  ;; CHECK-NEXT:   (local.get $ref)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $control-flow-in-set-value-safe (result i32)
+    ;; As above, but now the control flow in the value is safe: it does not
+    ;; escape out. We should optimize here.
+    (local $ref (ref null $struct))
+    (block $label
+      (struct.set $struct 0
+        (local.tee $ref
+          (struct.new $struct
+            (i32.const 1)
+          )
+        )
+        (if (result i32)
+          (i32.const 1)
+          (then
+            (i32.const 1337)  ;; the break to $out was replaced
+          )
+          (else
+            (i32.const 42)
+          )
+        )
+      )
+    )
+    (struct.get $struct 0
+      (local.get $ref)
+    )
+  )
+
+  ;; CHECK:      (func $control-flow-in-set-value-safe-call (type $4) (result i32)
+  ;; CHECK-NEXT:  (local $ref (ref null $struct))
+  ;; CHECK-NEXT:  (block
+  ;; CHECK-NEXT:   (local.set $ref
+  ;; CHECK-NEXT:    (struct.new $struct
+  ;; CHECK-NEXT:     (call $helper-i32
+  ;; CHECK-NEXT:      (i32.const 42)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (struct.get $struct 0
+  ;; CHECK-NEXT:   (local.get $ref)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $control-flow-in-set-value-safe-call (result i32)
+    ;; As above, but now the possible control flow is a call. It may throw, but
+    ;; that would go outside of the function, which is fine.
+    (local $ref (ref null $struct))
+    (block $label
+      (struct.set $struct 0
+        (local.tee $ref
+          (struct.new $struct
+            (i32.const 1)
+          )
+        )
+        (call $helper-i32 (i32.const 42))  ;; the if was replaced by this call
```

Comment:
We should check that this works with a `return`, too.

Reply:
A return would make the code unreachable, and this optimization would ignore it. Is that what you mean here?

Reply:
Ah, no, I had a conditional return in mind to match the conditional throw implicit in the call.

Reply:
We have a test with a `br` in an `if` - does that not cover the same as a `return` in an `if`? Or what do you mean by "conditional return"?

Reply:
I guess the return is a little different because it exits the function entirely, just like a `throw` or a `call`. In practice I would expect this to be covered by the same code paths as `call`, so in that sense it's not necessary.

Reply:
Now I follow you, thanks. Yeah, I guess return overlaps with call/break in a way that is slightly different than either, even if the two together seem to cover things. I added a testcase now, can't hurt to add more checks.